presto-main/src/main/java/com/facebook/presto/type/DecimalCasts.java

```diff
@@ -404,10 +429,12 @@ public static double shortDecimalToDouble(long decimal, long precision, long sca
     @UsedByGeneratedCode
     public static double longDecimalToDouble(Slice decimal, long precision, long scale, long tenToScale)
     {
-        // TODO: optimize
-        BigInteger decimalBigInteger = decodeUnscaledValue(decimal);
-        BigDecimal bigDecimal = new BigDecimal(decimalBigInteger, (int) scale);
-        return bigDecimal.doubleValue();
+        // If both decimal and scale can be represented exactly in double then compute rescaled and rounded result directly in double.
+        if (scale < DOUBLE_10_POW.length && compareUnsigned(decimal, MAX_EXACT_DOUBLE) <= 0) {
+            return (double) unscaledDecimalToUnscaledLongUnsafe(decimal) / DOUBLE_10_POW[(int) scale];
+        }
+
+        return parseDouble(Decimals.toString(decimal, (int) scale));
```

Comment:
add comment:

`TODO: optimize and convert directly to double in similar fashion as in double to decimal casts`